system/jlib/jthread.cpp

```diff
@@ -1299,8 +1334,42 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
         }
         return false;
     }
-};
+    void setPoolSize(unsigned newPoolSize)
+    {
+        CriticalBlock block(crit);
+        if (newPoolSize > defaultmax)
+        {
+            availsem.signal(newPoolSize - defaultmax);
+            defaultmax = newPoolSize;
+            return;
+        }
 
+        if (newPoolSize < defaultmax)
+        {
+            slotsToRelease += defaultmax - newPoolSize;
+            defaultmax = newPoolSize;
+            while (slotsToRelease)
+            {
+                if (!availsem.wait(0))
+                    break; // none available
+                --slotsToRelease;
+            }
+            if (slotsToRelease && runningCount())
```

Comment:
If while loop above failed to decrease by amount needed, then I think can only be true if some are running, so should assert.